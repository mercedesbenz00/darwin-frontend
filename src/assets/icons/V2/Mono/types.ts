export enum IconVariant {
  DEFAULT = 'default',
  DISABLED = 'disabled',
  SELECTED = 'selected',
  NEGATIVE = 'negative',
}
