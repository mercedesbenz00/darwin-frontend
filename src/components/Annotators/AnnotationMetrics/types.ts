import { MembershipPayload, DatasetAnnotationReportPayload } from '@/store/types'

export type AnnotationDataRange = 'total' | 'month' | 'week' | 'day'
export type AnnotationDataGranularity = 'day' | 'hour'

export type ChartDimension =
  'annotationTime' |
  'imagesAnnotated' |
  'avgTimePerAnnotation' |
  'annotationsCreated' |
  'reviewPassRate'

export type AdaptedDataPoint =
  {[k in Exclude<ChartDimension, 'reviewPassRate' | 'avgTimePerAnnotation'>]: number } &
  {[k in Extract<ChartDimension, 'reviewPassRate' | 'avgTimePerAnnotation'>]: number | null} &
  { userId: number | null, date: string }

export type AggregatedDataPoint = Pick<
    DatasetAnnotationReportPayload,
    | 'annotation_time'
    | 'annotations_approved'
    | 'annotations_created'
    | 'date'
    | 'images_annotated'
    | 'images_approved'
    | 'images_rejected'
    | 'user_id'
  >

export type AnnotationDimensionData = {
  annotationTime: number,
  annotationsCreated: number,
  avgTimePerAnnotation: number | null,
  imagesAnnotated: number,
  reviewPassRate: number | null
}

export type AnnotationChartData = {
  data: AdaptedDataPoint[]
  dimensions: AnnotationDimensionData
  member?: MembershipPayload
  visible: boolean
}

export type AnnotationBarChartData = {
  data: AggregatedDataPoint[]
  member?: MembershipPayload
}

/**
 * Format for a single item to be shown in a chart tooltip.
 *
 * A tooltip will contain multiple items
 */
export type ChartTooltipData = { member: MembershipPayload | null, value: string}
