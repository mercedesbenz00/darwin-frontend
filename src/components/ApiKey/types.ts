export type AbilityGroup =
  'manage_comments' |
  'manage_tasks' |
  'manage_datasets' |
  'manage_team' |
  'manage_team_members' |
  'manage_models'

export type Group = { id: AbilityGroup; name: string }
export type GroupOption = Group & { selected: boolean }

export enum SupportedLanguages {
  shell = 'shell',
  python = 'python',
  javascript = 'javascript',
  elixir = 'elixir',
  cli = 'cli'
}

export type AccessCodeExample = { language: SupportedLanguages, code: string }
