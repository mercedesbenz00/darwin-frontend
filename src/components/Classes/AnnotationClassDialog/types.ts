import { AnnotationClassPayload, AnnotationClassMetadata, AnnotationTypeName } from '@/store/types'

type PendingFields =
  | 'crop_key'
  | 'crop_url'
  | 'image_height'
  | 'image_width'
  | 'key'
  | 'original_image_url'
  | 'scale'
  | 'x'
  | 'y'

/**
 * Holds a class image that isn't necessarily persisted to the backend yet.
 * That means it might be missing any of the url fields
 *
 * It can be saved to backend once at least the `key` field is set, as that one
 * can be used to infer/compute all the other fields.
 *
 * - `crop_key` - repleace "original" with "crop" in `key` field
 * - `crop_url` - signed s3 url (by backend), for `crop_key`
 * - `original_image_url` - signed s3 url (by backend) for `key`
 */
export type PendingClassImage =
  Omit<AnnotationClassPayload['images'][0], PendingFields>
  & Partial<Pick<AnnotationClassPayload['images'][0], PendingFields>>

// eslint-disable-next-line camelcase
export type PendingClassImageWithUrlSet = PendingClassImage & { original_image_url: string }

export const isClassImage = (
  image: PendingClassImage
): image is AnnotationClassPayload['images'][0] =>
  !!image.key && !!image.crop_key

export type AnnotationClassData = {
  datasets: AnnotationClassPayload['datasets']
  description: string | null
  hotkey: null | string
  id: number | null
  images: PendingClassImage[]
  metadata: AnnotationClassMetadata
  name: string | null
}

export type DefaultAnnotationClassData = Partial<AnnotationClassData> & {
  selectedMainAnnotationType?: AnnotationTypeName
}

export type AnnotationClassValidationErrors = {
  annotationTypes?: string
  name?: string
  skeleton?: string
}

export const VIEWPORT_WIDTH: number = 150
export const VIEWPORT_HEIGHT: number = 150

export type CropInfo = {
  orientation: number
  points: [string, string, string, string]
  zoom: number
}

/**
 * VueCroppie is really archaic and requires us to get crop info via $ref,
 * so this type defines that ref
 */
export type Croppie = {
  /** Binds options to the underlying croppie library */
  bind: (options: Pick<CropInfo, 'points' | 'zoom'> & { url: string}) => void

  /** Retrieves current crop info */
  get: () => CropInfo
  result: (options?: Record<string, string>, callback?: Function) => void

  /** Sets scale/zoom level of cropper */
  setZoom: (value: number) => void
}
