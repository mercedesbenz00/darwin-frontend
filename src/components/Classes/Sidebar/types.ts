import { AnnotationTypeName } from '@/store/types'

export type AnnotationTypeFilterItemType = {
  id: string | number
  label: string
  icon?: string
  name: AnnotationTypeName
  color: string
  count: number | null
  [key: string]: any
}
