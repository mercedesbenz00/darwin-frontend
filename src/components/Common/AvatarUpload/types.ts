export type AvatarUploadData = {
  hash: string
  file: File
  type: string | null
}
