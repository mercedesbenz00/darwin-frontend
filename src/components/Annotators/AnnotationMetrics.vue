<template>
  <div class="metrics">
    <div class="metrics__header">
      <h2 class="metrics__header__left metrics__header__title">
        {{ dataset.name }}
      </h2>
      <div class="metrics__header__right">
        <tab-selector
          :options="dataRanges"
          :value="dataRange"
          @change="dataRange = $event"
        />
        <loading-wrapper
          :loading="downloadingReport"
          background-color="transparent"
          size="small"
          class="metrics__header__download"
        >
          <download-button @click="download" />
        </loading-wrapper>
      </div>
    </div>
    <div class="metrics__stats metrics__stats--top">
      <progress-donut-chart
        :dataset="dataset"
        subtext="Overall Progress"
        class="metrics__stats__donut-chart"
      />
      <div class="metrics__stats__bar-chart">
        <lazy-render>
          <annotation-bar-chart
            :data-range="dataRange"
            :dataset="dataset"
            :by-user-report="byUserReport"
            :total-report="totalReport"
          />
        </lazy-render>
      </div>
    </div>
    <lazy-render>
      <data-loader
        :dataset="dataset"
        :from="from"
        :data-range="dataRange"
      />
    </lazy-render>
    <annotator-stats
      class="metrics__stats"
      :by-user-report="byUserReport"
      :total-report="totalReport"
      :data-range="dataRange"
      :dataset="dataset"
      :data-granularity="dataGranularity"
    />
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'
import { State } from 'vuex-class'

import LazyRender from '@/components/Common/LazyRender'
import LoadingWrapper from '@/components/Common/LoadingWrapper.vue'
import TabSelector from '@/components/Common/TabSelector/TabSelector.vue'
import { TabSelectorOption } from '@/components/Common/TabSelector/types'
import { AnnotationReport } from '@/store/modules/annotators/types'
import { DatasetPayload, DatasetAnnotationReportPayload } from '@/store/types'

import AnnotationBarChart from './AnnotationMetrics/AnnotationBarChart.vue'
import AnnotatorStats from './AnnotationMetrics/AnnotatorStats.vue'
import DataLoader from './AnnotationMetrics/DataLoader'
import DownloadButton from './AnnotationMetrics/DownloadButton.vue'
import ProgressDonutChart from './AnnotationMetrics/ProgressDonutChart.vue'
import { AnnotationDataGranularity, AnnotationDataRange } from './AnnotationMetrics/types'
import { fromFromRange, granularityFromRange } from './AnnotationMetrics/utils'

const RANGE_OPTIONS: TabSelectorOption[] = [
  { value: 'total', description: 'All data from the beginning, with daily resolution' },
  { value: 'month', description: 'Data from the past 30 days, with daily resolution' },
  { value: 'week', description: 'Data from the past week, with hourly resolution' },
  { value: 'day', description: 'Data from the past 24 hours, with hourly resolution' }
]

@Component({
  name: 'annotation-metrics',
  components: {
    AnnotationBarChart,
    AnnotatorStats,
    DataLoader,
    DownloadButton,
    LazyRender,
    LoadingWrapper,
    ProgressDonutChart,
    TabSelector
  }
})
export default class AnnotationMetrics extends Vue {
  @Prop({ required: true, type: Object })
  dataset!: DatasetPayload

  dataRange: AnnotationDataRange = 'month'
  dataRanges = RANGE_OPTIONS

  get from (): string | null {
    return fromFromRange(this.dataRange)
  }

  get dataGranularity (): AnnotationDataGranularity {
    return granularityFromRange(this.dataRange)
  }

  // reports

  @State(state => state.annotators.annotationReports)
  annotationReports!: AnnotationReport[]

  get byUserReport (): DatasetAnnotationReportPayload[] {
    const { annotationReports: reports } = this

    const report = reports.find(r =>
      r.params.datasetId === this.dataset.id &&
      r.params.groupBy === 'dataset,user' &&
      r.params.granularity === this.dataGranularity &&
      r.params.from === this.from
    )

    return report ? report.data : []
  }

  get totalReport (): DatasetAnnotationReportPayload[] {
    const { annotationReports: reports } = this
    const report = reports.find(r =>
      r.params.datasetId === this.dataset.id &&
      r.params.groupBy === 'dataset' &&
      r.params.granularity === this.dataGranularity &&
      r.params.from === this.from
    )

    return report ? report.data : []
  }

  // actions

  downloadingReport: boolean = false

  async download () {
    this.downloadingReport = true

    const params = {
      dataset: this.dataset,
      granularity: this.dataGranularity
    }

    const { error } = await this.$store.dispatch('annotators/downloadAnnotationReport', params)
    if (error) { this.$store.dispatch('toast/warning', { content: error.message }) }

    this.downloadingReport = false
  }
}
</script>

<style lang="scss" scoped>
.metrics {
  @include col;
}

.metrics__header {
  -webkit-position: sticky;
  position: sticky;
  z-index: 1;
  top: 0;
  padding: 15px;
  background: $colorSecondaryLight3;
  align-items: center;

  @include row;

  margin-bottom: 15px;
}

.metrics__header__title {
  @include typography(xl, headers, bold);
}

.metrics__header__left {
  flex: auto;
}

.metrics__header__right {
  :not(:last-child) {
    margin-right: 25px;
  }
  @include row--center;
  flex: 1;

  justify-content: flex-end;
}

.metrics__header__download {
  width: 44px;
  height: 44px;
}

.metrics__stats {
  border-radius: 3px;
  width: 100%;
}

.metrics__stats:not(:last-child) {
  margin-bottom: 15px;
}

// dataset title and date range selection are also sticky, so we must override default
// sticky properties on the child component
.metrics__stats :deep(.stats__collapse-button) {
  top: 74px;
}

.metrics__stats :deep(.spinner) {
  padding: 15px;
}

.metrics__stats--top {
  display: grid;
  grid-template-columns: min-content auto;
}

.metrics__stats__donut-chart{
  width: 300px;
}

.metrics__stats__bar-chart{
  flex: 1;
}

</style>

<!-- eslint-disable vue-scoped-css/enforce-style-type -->
<style lang="scss">
.metrics__header__download {
  .v-loading {
    top: 7px !important;
  }
}
</style>
