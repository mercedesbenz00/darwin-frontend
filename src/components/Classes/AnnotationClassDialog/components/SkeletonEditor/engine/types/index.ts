export * from './SkeletonEdgeType'
export * from './SkeletonNodeType'
