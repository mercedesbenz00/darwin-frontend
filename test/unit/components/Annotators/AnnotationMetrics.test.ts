import { shallowMount, createLocalVue } from '@vue/test-utils'
import flushPromises from 'flush-promises'
import moment from 'moment'
import Vuex from 'vuex'

import createTestStore from 'test/unit/createTestStore'
import { buildDatasetAnnotationReportPayload, buildDatasetPayload } from 'test/unit/factories'

import AnnotationMetrics from '@/components/Annotators/AnnotationMetrics.vue'
import { DatasetAnnotationReportPayload } from '@/store/types'

const localVue = createLocalVue()
localVue.use(Vuex)

const sfh = buildDatasetPayload({ id: 1, name: 'SFH' })

const sfhAnnotationReport: DatasetAnnotationReportPayload[] = [
  buildDatasetAnnotationReportPayload({
    annotation_time: 10,
    annotations_created: 5,
    annotations_approved: 3
  }),
  buildDatasetAnnotationReportPayload({
    annotation_time: 5,
    annotations_created: 10,
    annotations_approved: 3
  })
]

let store: ReturnType<typeof createTestStore>
let defaultFrom: string
beforeEach(() => {
  // moment.utc is used within the component,
  // so we need to make sure it's stubbed to a fixed value
  const fakeNow = moment.utc('2020-02-01T00:00:00')
  jest.spyOn(moment, 'utc').mockReturnValue(fakeNow)
  defaultFrom = '2019-12-01T00:00:00.000000'
  store = createTestStore()
})

it('matches snapshot', () => {
  const propsData = { dataset: sfh }
  const wrapper = shallowMount(AnnotationMetrics, { localVue, propsData, store })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot when loaded', () => {
  store.commit('annotators/PUSH_ANNOTATION_REPORT', {
    params: { datasetId: sfh.id, granularity: 'day', groupBy: 'dataset,user', from: defaultFrom },
    data: sfhAnnotationReport
  })

  store.commit('annotators/PUSH_ANNOTATION_REPORT', {
    params: { datasetId: sfh.id, granularity: 'day', groupBy: 'dataset', from: defaultFrom },
    data: sfhAnnotationReport
  })
  const propsData = { dataset: sfh }
  const wrapper = shallowMount(AnnotationMetrics, { localVue, propsData, store })
  expect(wrapper).toMatchSnapshot()
})

describe('report download', () => {
  it('downloads report with correct params', () => {
    const propsData = { dataset: sfh }
    const wrapper = shallowMount(AnnotationMetrics, { localVue, propsData, store })

    const dispatch = store.dispatch as jest.Mock
    dispatch.mockClear()

    wrapper.find('download-button-stub').vm.$emit('click')
    expect(dispatch).toHaveBeenCalledWith('annotators/downloadAnnotationReport', { dataset: sfh, granularity: 'day' })
    dispatch.mockClear()

    wrapper.setData({ dataRange: 'month' })
    wrapper.find('download-button-stub').vm.$emit('click')
    expect(dispatch).toHaveBeenCalledWith('annotators/downloadAnnotationReport', { dataset: sfh, granularity: 'day' })
    dispatch.mockClear()

    wrapper.setData({ dataRange: 'week' })
    wrapper.find('download-button-stub').vm.$emit('click')
    expect(dispatch).toHaveBeenCalledWith('annotators/downloadAnnotationReport', { dataset: sfh, granularity: 'hour' })
    dispatch.mockClear()

    wrapper.setData({ dataRange: 'day' })
    wrapper.find('download-button-stub').vm.$emit('click')
    expect(dispatch).toHaveBeenCalledWith('annotators/downloadAnnotationReport', { dataset: sfh, granularity: 'hour' })
  })

  it('shows toast error on download failure', async () => {
    jest.spyOn(store, 'dispatch').mockResolvedValue({ error: { message: 'Fake error' } })
    const propsData = { dataset: sfh }
    const wrapper = shallowMount(AnnotationMetrics, { localVue, propsData, store })

    wrapper.find('download-button-stub').vm.$emit('click')

    await flushPromises()
    expect(store.dispatch).toHaveBeenCalledWith('toast/warning', { content: 'Fake error' })
  })
})
