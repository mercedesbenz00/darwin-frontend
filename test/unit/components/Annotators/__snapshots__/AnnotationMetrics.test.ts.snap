// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div class="metrics">
  <div class="metrics__header">
    <h2 class="metrics__header__left metrics__header__title">
      SFH
    </h2>
    <div class="metrics__header__right">
      <tab-selector-stub value="month" options="[object Object],[object Object],[object Object],[object Object]"></tab-selector-stub>
      <loading-wrapper-stub backgroundcolor="transparent" size="small" theme="grey" class="metrics__header__download">
        <download-button-stub></download-button-stub>
      </loading-wrapper-stub>
    </div>
  </div>
  <div class="metrics__stats metrics__stats--top">
    <progress-donut-chart-stub dataset="[object Object]" subtext="Overall Progress" class="metrics__stats__donut-chart"></progress-donut-chart-stub>
    <div class="metrics__stats__bar-chart">
      <lazy-render-stub>
        <annotation-bar-chart-stub byuserreport="" totalreport="" datarange="month" dataset="[object Object]"></annotation-bar-chart-stub>
      </lazy-render-stub>
    </div>
  </div>
  <lazy-render-stub>
    <data-loader-stub dataset="[object Object]" datarange="month" from="2020-01-01T00:00:00.000000"></data-loader-stub>
  </lazy-render-stub>
  <annotator-stats-stub byuserreport="" totalreport="" datarange="month" dataset="[object Object]" datagranularity="day" class="metrics__stats"></annotator-stats-stub>
</div>
`;

exports[`matches snapshot when loaded 1`] = `
<div class="metrics">
  <div class="metrics__header">
    <h2 class="metrics__header__left metrics__header__title">
      SFH
    </h2>
    <div class="metrics__header__right">
      <tab-selector-stub value="month" options="[object Object],[object Object],[object Object],[object Object]"></tab-selector-stub>
      <loading-wrapper-stub backgroundcolor="transparent" size="small" theme="grey" class="metrics__header__download">
        <download-button-stub></download-button-stub>
      </loading-wrapper-stub>
    </div>
  </div>
  <div class="metrics__stats metrics__stats--top">
    <progress-donut-chart-stub dataset="[object Object]" subtext="Overall Progress" class="metrics__stats__donut-chart"></progress-donut-chart-stub>
    <div class="metrics__stats__bar-chart">
      <lazy-render-stub>
        <annotation-bar-chart-stub byuserreport="[object Object],[object Object]" totalreport="[object Object],[object Object]" datarange="month" dataset="[object Object]"></annotation-bar-chart-stub>
      </lazy-render-stub>
    </div>
  </div>
  <lazy-render-stub>
    <data-loader-stub dataset="[object Object]" datarange="month" from="2019-12-01T00:00:00.000000"></data-loader-stub>
  </lazy-render-stub>
  <annotator-stats-stub byuserreport="[object Object],[object Object]" totalreport="[object Object],[object Object]" datarange="month" dataset="[object Object]" datagranularity="day" class="metrics__stats"></annotator-stats-stub>
</div>
`;
